tempesta_fw/connection.c

```diff
@@ -121,7 +132,17 @@ tfw_connection_destruct(TfwConnection *conn)
 void
 tfw_connection_send(TfwConnection *conn, TfwMsg *msg)
 {
-	ss_send(conn->sk, &msg->skb_list);
+	struct sock *sk;
+
+	spin_lock(&conn->splock);
+	sk = conn->sk;
+	if (sk)
+		ss_sock_hold(sk);
+	spin_unlock(&conn->splock);
+	if (sk) {
+		ss_send(sk, &msg->skb_list);
+		sock_put(sk);
```

Comment:
Looks bit unusual bedause double sk checking on most probable workflow, I'd rewrite it like

        sk = conn->sk;
        if (unlikely(!sk)) {
                spin_unlock(&conn->splock);                                    
                return;                                                        
        }                                                                      
        
        ss_sock_hold(sk);
        spin_unlock(&conn->splock);                                            
        
        ss_send(sk, &msg->skb_list);                                           
        
        ss_sock_put(sk);